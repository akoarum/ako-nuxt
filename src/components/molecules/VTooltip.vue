<template>
  <div class="vTooltip">
    <button
      ref="btn"
      :aria-controls="`balloon-${componentId}`"
      :aria-expanded="String(isVisible)"
      type="button"
      class="vTooltip__btn"
      @click="updateVisible(true)"
    >
      <span>{{ label }}</span>
    </button>
    <FadeTransition>
      <div v-show="isVisible">
        <VBalloon
          :id="`balloon-${componentId}`"
          ref="content"
          :aria-hidden="String(!isVisible)"
          :tabindex="isVisible ? '0' : '-1'"
          :style="contentStyled"
          class="vTooltip__content"
        >
          <slot />
        </VBalloon>
        <VMask
          class="vTooltip__mask"
          transparent
          @click="updateVisible(false)"
        />
      </div>
    </FadeTransition>
  </div>
</template>

<script>
import WindowSize from '~/utils/WindowSize'
import { FadeTransition } from '~/components/presenter'
import { VBalloon, VMask } from '~/components/atoms'

export default {
  components: { FadeTransition, VBalloon, VMask },
  props: {
    label: { type: String, required: true }
  },
  data() {
    return {
      isVisible: false,
      windowSize: new WindowSize().initialize(),
      contentSize: {
        width: 0,
        height: 0
      },
      btnRect: {
        x: 0,
        y: 0,
        width: 0,
        height: 0
      },
      resizeTime: null
    }
  },
  computed: {
    componentId() {
      return this._uid
    },
    windowWidth() {
      return this.windowSize.width
    },
    contentStyled() {
      let styles = {}

      styles = Object.assign({}, styles, this.leftPositionStyled)
      styles = Object.assign({}, styles, this.verticalPositionStyled)

      return styles
    },
    basePosition() {
      return {
        x: this.btnRect.x + this.btnRect.width / 2,
        y: this.btnRect.y
      }
    },
    leftPositionStyled() {
      const styles = {}

      if (this.basePosition.x - this.contentSize.width / 2 > 0) {
        if (
          this.basePosition.x + this.contentSize.width / 2 >
          this.windowWidth
        ) {
          const position =
            this.windowWidth - this.contentSize.width - this.btnRect.x
          styles.left = `${position}px`
        } else {
          styles.left = '50%'
          styles.transform = 'translateX(-50%)'
        }
      } else {
        styles.left = `-${this.btnRect.x}px`
      }

      return styles
    },
    verticalPositionStyled() {
      const styles = {}

      if (this.basePosition.y - this.contentSize.height - 14 > 0) {
        styles.top = `-${this.contentSize.height + 14}px`
      } else {
        styles.top = 'calc(100% + 14px)'
      }

      return styles
    }
  },
  watch: {
    isVisible(newValue) {
      if (newValue) {
        window.addEventListener('resize', this.handleResize)
        this.$nextTick(() => {
          this.updatePositionData()
        })
      } else {
        window.removeEventListener('resize', this.handleResize)
      }
    }
  },
  methods: {
    handleResize() {
      clearTimeout(this.resizeTime)
      this.resizeTime = setTimeout(() => {
        this.updatePositionData()
      }, 80)
    },
    updateVisible(newValue) {
      this.isVisible = newValue
    },
    updatePositionData() {
      if (!this.$refs.btn) return false
      const btnRect = this.$refs.btn.getBoundingClientRect()
      this.contentSize.width = this.$refs.content.$el.clientWidth
      this.contentSize.height = this.$refs.content.$el.clientHeight
      this.btnRect.x = btnRect.x
      this.btnRect.y = btnRect.y
      this.btnRect.width = btnRect.width
      this.btnRect.height = btnRect.height
    }
  }
}
</script>

<style lang="scss" scoped>
.vTooltip {
  position: relative;
  display: inline-block;
}
.vTooltip__content {
  position: absolute;
  z-index: 11;
  transition: left 0.3s $easing;
}
.vTooltip__btn {
  display: inline-block;
  border-radius: 2px;
  background: $color-base;
  font-weight: bold;
  line-height: 1;
  color: #fff;

  @include media-min {
    padding: 5px;
    font-size: 1.4rem;
  }

  @include media-max {
    padding: 3px;
    font-size: 1.5rem;
  }
}
.vTooltip__text {
  white-space: nowrap;
}
.vTooltip__mask {
  position: fixed;
  left: 0;
  top: 0;
  z-index: 10;
  width: 100%;
  height: 100%;
}
</style>
