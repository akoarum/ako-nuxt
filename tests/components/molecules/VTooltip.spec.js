import { shallowMount } from '@vue/test-utils'
import VTooltip from '~/components/molecules/VTooltip'

describe('VTooltip', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(VTooltip, {
      propsData: {
        label: '？'
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('computed', () => {
    it('componentId: 一意のIDであるuidを返す', () => {
      expect(VTooltip.computed.componentId.call({ _uid: 100 })).toBe(100)
    })

    it('contentStyled: バルーン用のスタイルを返す', () => {
      const calledObject = {
        leftPositionStyled: {
          left: '-50px'
        },
        verticalPositionStyled: {
          top: '-50px'
        }
      }
      expect(VTooltip.computed.contentStyled.call(calledObject)).toEqual({
        left: '-50px',
        top: '-50px'
      })
    })

    it('basePosition: ボタンのトップ、横中央を返す', () => {
      const calledObject = {
        btnRect: {
          width: 30,
          x: 300,
          y: 50
        }
      }
      expect(VTooltip.computed.basePosition.call(calledObject)).toEqual({
        x: 315,
        y: 50
      })
    })

    describe('leftPositionStyled: バルーンのスタイル left を返す', () => {
      it('ボタンに対してセンタリングで表示できればセンタリングにする', () => {
        const calledObject = {
          btnRect: {
            width: 30,
            x: 300
          },
          contentSize: {
            width: 100
          },
          basePosition: {
            x: 315
          }
        }
        expect(VTooltip.computed.leftPositionStyled.call(calledObject)).toEqual({
          left: '50%',
          transform: 'translateX(-50%)'
        })
      })

      it('センタリングだと左に突き抜ける場合はブラウザ内に収まる位置にする', () => {
        const calledObject = {
          btnRect: {
            width: 30,
            x: 10
          },
          contentSize: {
            width: 100
          },
          basePosition: {
            x: 45
          }
        }
        expect(VTooltip.computed.leftPositionStyled.call(calledObject)).toEqual({
          left: '-10px'
        })
      })

      it('センタリングだと右に突き抜ける場合はブラウザ内に収まる位置にする', () => {
        const calledObject = {
          btnRect: {
            width: 30,
            x: 750
          },
          contentSize: {
            width: 100
          },
          basePosition: {
            x: 765
          },
          windowWidth: 780
        }
        expect(VTooltip.computed.leftPositionStyled.call(calledObject)).toEqual({
          left: '-70px'
        })
      })
    })

    describe('verticalPositionStyled: 上下の位置を指定する', () => {
      it('ボタンの上に表示できる場合は上にする', () => {
        const calledObject = {
          contentSize: {
            height: 50
          },
          basePosition: {
            y: 100
          }
        }
        expect(VTooltip.computed.verticalPositionStyled.call(calledObject)).toEqual({
          top: '-64px'
        })
      })

      it('上に収まらない場合は下にする', () => {
        const calledObject = {
          contentSize: {
            height: 50
          },
          basePosition: {
            y: 30
          }
        }
        expect(VTooltip.computed.verticalPositionStyled.call(calledObject)).toEqual({
          top: 'calc(100% + 14px)'
        })
      })
    })
  })

  describe('methods', () => {
    describe('updateVisible', () => {
      it('引数をisVisibleに反映する', () => {
        wrapper.vm.updateVisible(true)
        expect(wrapper.vm.isVisible).toBe(true)

        wrapper.vm.updateVisible(false)
        expect(wrapper.vm.isVisible).toBe(false)
      })
    })
  })
})
