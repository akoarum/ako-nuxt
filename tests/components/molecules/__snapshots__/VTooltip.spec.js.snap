// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTooltip initialized snapshot 1`] = `
<div
  class="vTooltip"
>
  <button
    aria-controls="balloon-6"
    aria-expanded="false"
    class="vTooltip__btn"
    type="button"
  >
    <span>
      ï¼Ÿ
    </span>
  </button>
   
  <fadetransition-stub>
    <div
      style="display: none;"
    >
      <vballoon-stub
        aria-hidden="true"
        class="vTooltip__content"
        id="balloon-6"
        style="left: -0px;"
        tabindex="-1"
      />
       
      <vmask-stub
        class="vTooltip__mask"
        transparent="true"
      />
    </div>
  </fadetransition-stub>
</div>
`;
